### PR TITLE
80 colidable system

### PR DESCRIPTION
**Pull request by 31Nathan**
Adding a collidable system
Maximum optimization to not check twice the same entities

Need to talk about the way the ordinate and the abscissa will be calculated (if it's like the SFML way then that's done)

### CODE REVIEW

File: tests/unit_tests/libs_tests/ECS_tests/Systems_tests/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,7 @@ set (SRC_ROOT ${PROJECT_SOURCE_DIR}/tests/unit_tests/libs_tests/ECS_tests/System
 
 set(TEST_SRC_ROOT
     ${SRC_ROOT}/Movement_System_tests.cpp
+    ${SRC_ROOT}/CollidableSystemTest.cpp
```
Comment (LucasTesnier) [inline tests/unit_tests/libs_tests/ECS_tests/Systems_tests/CMakeLists.txt:8]: match you file case

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Some recommendation, otherwise very good job

**Review by LucasTesnier — APPROVED** [review]

File: libs/GameSystems/CollidableSystem.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#ifndef COLLIDABLESYSTEM_HPP_
+#define COLLIDABLESYSTEM_HPP_
+
+#include "World/World.hpp"
+
+namespace ecs
+{
+    /// @brief This is the Collide System.
+    /// This system will check every entities, see if they collide and apply damage on consequence
+    /// Needed Component : Collidable, Position, Size, Damage
```
Comment (TomDUVAL-MAHE) [inline libs/GameSystems/CollidableSystem.hpp:17]: Good !

**Review by TomDUVAL-MAHE — DISMISSED** [review]

File: libs/GameSystems/CollidableSystem.hpp (hunk 2 above)
Comment (ZQUEMA) [reply libs/GameSystems/CollidableSystem.hpp:17]: I like It

File: libs/GameSystems/CollidableSystem.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    struct Collide : public System {
+        /// @brief isSameWidth Function
+        /// This function checks if two entities have a common part on the map's width
+        /// @param fstPos position of the first Entity
+        /// @param sndPos position of the second Entity
+        /// @param fstSize size of the first Entity
+        /// @param sndSize size of the second Entity
+        bool isSameWidth(Position &fstPos, Position &sndPos, Size &fstSize, Size &sndSize)
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:25]: Please put it in a source (.cpp) file or add the inline mention

File: libs/GameSystems/CollidableSystem.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            return fstBelowSnd || fstAboveSnd || sndBelowFst || snfAboveFst;
+        }
+
+        /// @brief isSameHeight Function
+        /// This function checks if two entities have a common part on the map's height
+        /// @param fstPos position of the first Entity
+        /// @param sndPos position of the second Entity
+        /// @param fstSize size of the first Entity
+        /// @param sndSize size of the second Entity
+        bool isSameHeight(Position &fstPos, Position &sndPos, Size &fstSize, Size &sndSize)
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:41]: Please put it in a source (.cpp) file or add the inline mention

File: libs/GameSystems/CollidableSystem.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            bool sndToTheLeftFst = sndPos.x <= fstPos.x + fstSize.x && fstPos.x + fstSize.x <= sndPos.x + sndSize.x;
+
+            return fstToTheRightSnd || fstToTheLeftSnd || sndToTheRightFst || sndToTheLeftFst;
+        }
+
+        /// @brief Collide Function
+        /// This function checks all the given entities passed in params, to see if they collide
+        /// @param fstEntities first vector of entities to be checked
+        /// @param sndEntities second vector of Entities to be checked
+        void collide(
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:55]: Please put it in a source (.cpp) file or add the inline mention

File: libs/GameSystems/CollidableSystem.hpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        /// @brief Collide Function
+        /// This function checks all the given entities given in params, to see if they collide
+        /// @param fstEntities first vector of entities to be checked
+        /// @param sndEntities second vector of Entities to be checked
+        void collide(
+            std::vector<std::shared_ptr<ecs::Entity>> &fstEntities, std::vector<std::shared_ptr<ecs::Entity>> &sndEntities)
+        {
+            for (std::shared_ptr<ecs::Entity> fstEntity : fstEntities) {
```
[the remark this answers is not in the document]
Comment (ZQUEMA) [reply libs/GameSystems/CollidableSystem.hpp:58]: Nope you can directly precise the type and it's the best option

File: libs/GameSystems/CollidableSystem.hpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            collide(enemiesProjectiles, obstacles);
+        }
+
+        /// @brief AllyCollide Function
+        /// Check for every alliedEntities if they collide with enemyEntities
+        /// @param allyEntities vector of entities which need to be checked if there is collision with the following params
+        /// @param enemies vector of entities containing the enemies
+        /// @param enemiesProjectiles vector of entities containing the enemies' projectiles
+        /// @param obstacles vector of entities containing the obstacles
+        void allyCollide(std::vector<std::shared_ptr<ecs::Entity>> &allyEntities,
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:100]: Please put it in a source (.cpp) file or add the inline mention

File: libs/GameSystems/CollidableSystem.hpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                    }
+                }
+            }
+        }
+
+        /// @brief enemyCollide Function
+        /// Check for enemy entities if they collide with obstacles
+        /// @param enemiesProjectiles vector of entities containing the enemies' projectiles
+        /// @param obstacles vector of entities containing the obstacles
+        void enemyCollide(std::vector<std::shared_ptr<ecs::Entity>> &enemiesProjectiles,
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:88]: Please put it in a source (.cpp) file or add the inline mention

File: libs/GameSystems/CollidableSystem.hpp [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            std::vector<std::shared_ptr<ecs::Entity>> &enemies,
+            std::vector<std::shared_ptr<ecs::Entity>> &enemiesProjectiles,
+            std::vector<std::shared_ptr<ecs::Entity>> &obstacles)
+        {
+            collide(allyEntities, enemies);
+            collide(allyEntities, enemiesProjectiles);
+            collide(allyEntities, obstacles);
+        }
+
+        inline void run(World &world) override final
```
Comment (ZQUEMA) [inline libs/GameSystems/CollidableSystem.hpp:110]: Please put it in a source (.cpp) file

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
The logic is good but some functions should be move in a source file

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Non inline function -> CPP

**Review by LucasTesnier — DISMISSED** [review]
NTR!

**Review by LucasTesnier — APPROVED** [review]

**Review by ZQUEMA — APPROVED** [review]

**Pull request merged**